Draft: Create Network driver instead of directly using sockets.

Jakio815 opened the pull request:
lf-lang/lingua-franca#2256
# Open problems
[] OPENSSL_init_crypto(OPENSSL_INIT_NO_ATEXIT, NULL); - the "atexit" problem.

# Netdriver
Current LF code of the reactor-c uses TCP and UDP sockets directly, limiting to scale network options. This PR introduces a network driver to support various network types.
```
typedef struct netdrv_t {
  void* priv;
  unsigned int read_remaining_bytes;
  int federate_id;
} netdrv_t;
```
The `void* priv` will be casted depending on the network type.

# APIs.
The main design is as a client and server. The RTI is always a server. Federates can be servers and clients. As a client, it should be able to connect to the RTI, and also connect to other federates. As a server, it should be able to accept connections from other federates for two reasons: decentralized coordination and physical connections.
So, the main design follows TCP socket server-client design. 
The RTI will have a netdriver that accepts other federate connections, and a netdriver will be allocated for each federate.
The federate will have a netdriver to connect to the RTI as a client.
Next, a netdriver to accept other federate's connection as a server.
Finally netdrivers to communicate with each inbound and outbound p2p connections.

The description below separates the APIs as functions for clients, servers, and common each.
## Common 
### netdrv_t* initialize_netdrv();
Initializes netdriver, and allocates memory.
### void close_netdrv(netdrv_t* drv);
Frees memory allocated by netdriver.
### int write_to_netdrv(netdrv_t* drv, size_t num_bytes, unsigned char* buffer);
This sends the buffer to the netdriver.
TCP: write()
MQTT: publish()
SST: encapsulate message with SST, and send.

The below two each has additional error handling.
`int write_to_netdrv_close_on_error(netdrv_t* drv, size_t num_bytes, unsigned char* buffer);`
`void write_to_netdrv_fail_on_error(netdrv_t* drv, size_t num_bytes, unsigned char* buffer, lf_mutex_t* mutex,char* format, ...);`
### ssize_t read_from_netdrv(netdrv_t* drv, unsigned char* buffer, size_t buffer_length);
`ssize_t read_from_netdrv_close_on_error(netdrv_t* drv, unsigned char* buffer, size_t buffer_length);`
`void read_from_netdrv_fail_on_error(netdrv_t* drv, unsigned char* buffer, size_t buffer_length, lf_mutex_t* mutex, char* format, ...);`

## Server
### int create_server(netdrv_t* drv, int server_type, uint16_t port);
TCP: creates server socket, binds, and listens to server socket.
MQTT: Connects to broker. Subscribe to Topic “{Federation_ID}_RTI”
SST: creates server socket, init_SST(): loads keys

### netdrv_t* establish_communication_session(netdrv_t* netdrv);
TCP: accept() for federate socket requests, return fed_socket.
Mqtt:
1. (RTI is already subscribing to “{Federation_ID}_RTI”)
2. (Each federate publishes fed_id to “{Federation_ID}_RTI”)
3. RTI subscribes to “{Federation_ID}_fed{fed_id}_to_RTI“
4. (Federate subscribes to ”{Federation_ID}_RTI_to_fed{fed_id}”)
* Now there are two one-way channels for each RTI-fed connection.

SST: accept() sockets. RTI does handshake with federates. SST_session_ctx is created for each federate.

## Client
### int create_client(netdrv_t* drv, int federate_id);
TCP: Creates a real-time TCP socket.
MQTT: Federate connects to broker.
SST: init_SST().

### int connect_to_netdrv(netdrv_t* drv);
TCP: connect()
MQTT: Federate publishes to “{Federation_ID}_RTI”, then subscribes to ”{Federation_ID}_RTI_to_fed{fed_id}”
SST: Requests session key from Auth, and request connection to RTI or federate server.


# Major changes
## MQTT

| | |**\_RTI**| **\_RTI\_to\_fedID** | **\_fedID\_to\_RTI** | **\_fedServID** | **\_fedServID\_to\_fedCliID** | **\_fedCliID\_to\_fedServID** |
|:----------:|:----------:|:----------:|:----------:|:----------:|:----------:|:----------:|:----------:|
|RTI           |rti\_netdrv|sub           |                |                |                |                |                |
|                |fed\_netdrv|              |pub          |sub          |                 |                |                | 
|Fed         |rti\_netdrv|pub          |sub           |pub         |                |                 |               |
|                |serv\_netdrv|             |                |                |sub           |pub          |sub          |
|               |cli\_netdrv |                |                |                |pub          |sub          |pub          |



## SST
Netdriver supports SST.
```
reactor C {
  comm-type:SST
}
```
RTI must be built with `-DCOMM_TYPE=SST`.
### Instructions
1. Install sst-c-api
OpenSSL 3.0 above must be installed.
```
$ git clone https://github.com/iotauth/sst-c-api.git
$ cd sst-c-api && mkdir build && cd build
$ cmake ../
$ make
$ sudo make install
```
2. Generate keys and execute `Auth`.
Java 11 or above must be installed.
```
$ git clone https://github.com/iotauth/iotauth.git
$ cd iotauth/examples
$ ./generateAll.sh -g configs/lf_sst.graph 
$ cd ../auth/auth-server/
$ java -jar target/auth-server-jar-with-dependencies.jar -p ../properties/exampleAuth101.properties
```
The path must be in the same directory with `lingua-franca` for the script to work. This is for key paths, and auto generated `.config` files.
```
.
├── iotauth
├── lingua-franca
```
3. Install RTI with `-DCOMM_TYPE=SST`.
4. Add `comm-type:SST` to `.lf` code.
### Input for configuration is required.
Each federate binaries and the RTI needs a config as an input with the `-sst` or `--sst` option.

The code generator copies the skeleton config from `reactor-c/core/federated/network/SSTskeleton.config`.
It writes three components.
- entityInfo.name
- authInfo.pubkey.path
- entityInfo.privkey.path

The generated federate's sst config will be saved such as `test/C/fed-gen/SimpleFederated/src-gen/federate__fed1/core/federated/network/federate__fed1.config`

The RTI's config will be saved such as `test/C/fed-gen/SimpleFederated/src-gen/rti.config`. However, this RTI config is only for local environments.
```
./federate__fed1 -sst federate__fed1.config
./RTI -sst rti.config
```

# Refactoring
### federate.c
`lf_create_server()`: 
Separate the part sending the address_advertisement message to the RTI. The implementation is in each lf_STACK_support.c. In TCP connections, it will send the port as it did. In MQTT connections, it will do nothing, because it does not need the RTI to connect to another federate. (The MQTT will do a handshake using the federateID, which is already known).

`lf_connect_to_federate()`:
This is a very large function having three parts.
1-1. Send the RTI an ADDRESS_QUERY message, sending the remote federate's ID which it wants to connect to.
1-2. Receive ADDRESS_QUERY_REPLY message from RTI, getting the port number and IP address of the requested federate.
2. Connect to the remote federate, using the port and IP address.
3-1. Send my federateID and federationID
3-2. Receive an ACK.

I splitted this into three parts, using helper functions. However, part 1, sending address query messages, will have different implementations depending on the network communication type. So, it is defined in netdriver.h, and implemented in lf_STACK_support.c

### Maximum length of Federation ID is needed.
### Change response message of RTI to federate MSG_TYPE_ADDRESS_QUERY
 ** THIS IS ALREADY FIXED IN #392 
The federate first sends MSG_TYPE_ADDRESS_QUERY with total 5bytes (1 Header + 2 federate id).
Then, the RTI responds with the same header MSG_TYPE_ADDRESS_QUERY with (1 Header + 4 port_num + 4 IP address).

Due to my approach needing unique message headers, the response of the RTI will be MSG_TYPE_ADDRESS_QUERY_REPLY.

edwardalee reviewed (commented):
I realize this is just an early draft, but this looks like a good start, working in the right direction.  A couple of suggestions that may save time later:

1. You may want to base your changes on the federated-cleanup branch of reactor-c, which will be merged soon, I think.  It includes quite a few changes to the networking code.
2. Our coding style uses spaces only for indentation, no tabs.
3. Our coding style puts opening brackets on the same line and always encloses if statements in brackets when a new line is involved.  E.g., instead of:

```
int  netdrv_open(netdrv_t *drv)
{
	if (!drv)
		return -1;
	return drv->open(drv);
}
```
it should be
```
int  netdrv_open(netdrv_t *drv) {
    if (!drv) {
        return -1;
    }
    return drv->open(drv);
}
```
or
```
int  netdrv_open(netdrv_t *drv) {
    if (!drv) return -1;
    return drv->open(drv);
}
```

4. Header files should always document functions using JavaDoc style documentation.

Jakio815 commented:
@edwardalee Thank you for having a look. I'll keep in my what you mentioned. Especially for the indents and coding style, it was just the auto formatted. I'll fix it. Also I'll add up the documentation for header files. Thank you.

Jakio815 committed:
Try fixing net function for git history

Jakio815 committed:
Remove net_wait_for_federates for history

Jakio815 committed:
Fix function names to original names

Jakio815 committed:
Fix send_rejects

Jakio815 committed:
Fix respond_to_erroneous_connections

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Minor fix

Jakio815 committed:
Minor fix

Jakio815 committed:
Minor fix

Jakio815 committed:
Temporarily add create_real_time_tcp_socket_errexit()

Jakio815 committed:
Fix clock_synchronization thread

Jakio815 committed:
Add breaks to switch case

Jakio815 committed:
Handle_timed message

Jakio815 committed:
Fix handle adress query and address ad

Jakio815 committed:
Fix debug message

Jakio815 committed:
Fix read_netdrv to make it stateful

Jakio815 committed:
Fixed tagged message bugs

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Fix handle_federate_resign && fix comments && add message_type understand HMAC authentication.

Jakio815 committed:
Fix send_failed signal && add resign msg_type

Jakio815 committed:
Fix authenticated bugs

Jakio815 committed:
Erase unused code

Jakio815 committed:
Add timed message forwarding large buffers

Jakio815 committed:
Fix return of read messages to return the bytes read, not success and fail values.

Jakio815 committed:
Change read return type to ssize_t

Jakio815 committed:
Make inner switch case to be handled in a different helper function.

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Make socket_open to api

Jakio815 committed:
Add some comments

Jakio815 committed:
Fix netdrv_accept fixing /DistributedPhysicalActionUpstream.lf

Jakio815 committed:
Add commits

Jakio815 committed:
Add helper functions

Jakio815 committed:
Starting changes to federate.c

Jakio815 committed:
Finished until lf_create_server()

Jakio815 committed:
Finished lf_handle_p2p_connections_from_federates

Jakio815 committed:
Minor fix

Jakio815 committed:
Finished lf_connect_to_federate

Jakio815 committed:
Fix handling messsages

Jakio815 committed:
Fix federate's p2p tagged message and similiar messages

Jakio815 committed:
Fix socket open and netdrv free

Jakio815 committed:
Minor fix

Jakio815 committed:
Fix bug on tagged message

Jakio815 committed:
Fix clock sync

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Fix proto

byeonggiljun committed:
Return -1 when priv is not initialized yet and try to get the port number

Jakio815 committed:
Add ADDRESS_QUERY_REPLY msg

Jakio815 committed:
Merge branch 'netdriver' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Fix handle address query when port is unavailable.

Jakio815 commented:
> The reason for the error in [handle_address_query](https://github.com/lf-lang/reactor-c/blob/5ead7cc5055a42b20875069618dbe7fe24613362/core/federated/RTI/rti_remote.c#L593) is that [`get_port`](https://github.com/lf-lang/reactor-c/blob/5ead7cc5055a42b20875069618dbe7fe24613362/core/federated/RTI/rti_remote.c#L616) invokes
> 
> https://github.com/lf-lang/reactor-c/blob/9d2c3d574e27cd72d96aac8a32ae7636e87bb173/core/federated/network/lf_socket_support.c#L29
> 
> . So I made [`get_port`](https://github.com/lf-lang/reactor-c/blob/5ead7cc5055a42b20875069618dbe7fe24613362/core/federated/RTI/rti_remote.c#L616) returns `-1` if the socket is not initialized.
> Yet, the below line reports a segmentation fault now. @Jakio815 could you modify this part?
> 
> https://github.com/lf-lang/reactor-c/blob/5ead7cc5055a42b20875069618dbe7fe24613362/core/federated/RTI/rti_remote.c#L617

Thanks for your help. It seems that now the weird accept() failing does not appear. Still don't know why it failed at that point, but really thanks for your help!

Jakio815 committed:
Fix proto initializztion

Jakio815 committed:
Fix probe message

Jakio815 committed:
Remove unnecessary socket related structs

Jakio815 committed:
Minor changes

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Add include to clock_sync.h

Jakio815 committed:
Fix stack smashing for RTI

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Run clang-format

Jakio815 committed:
Fix wrong emerged

Jakio815 committed:
Fix merge conflict error

Jakio815 committed:
Finally fixed bug.... The federation ID should have a maximum value.

Jakio815 committed:
Fix hmac authentication MSG read.

Jakio815 committed:
Minor fix of comments

Jakio815 committed:
Add gitmodules adding sst-c-api

Jakio815 committed:
Add sst-c-api as submodule

Jakio815 committed:
Add lf_sst_support.c .h bases

Jakio815 committed:
Revert mistakes

Jakio815 committed:
Add empty mqtt support .c .h

Jakio815 committed:
Remove unneeded functions yet for sst_support.c

Jakio815 committed:
Clean net_util.h

Jakio815 committed:
Clean net_common.h

Jakio815 committed:
Clean rti_remote.h

Jakio815 committed:
Make socket_common. h .c && Implement create_server(), which is used for both fed and RTI servers.

Jakio815 committed:
Apply create_server() to rti and fed.

Jakio815 committed:
Minor fix

Jakio815 committed:
Cleanup socket_support.c & .h

Jakio815 committed:
Cleanup net_util .h .c

Jakio815 committed:
Update accept_connection to establilsh_communication_session

Jakio815 committed:
Add skeleton for lf_mqtt_support.c

Jakio815 committed:
Change all netdrv_accepts to establish_communication_session

Jakio815 committed:
Fix comments

Jakio815 committed:
Fix not to use clock_netdrv on rti.

Jakio815 committed:
Fix clock_synchronization_thread

Jakio815 committed:
Minor cleanup

chanijjani committed:
Enable MQTT build

chanijjani committed:
Merge branch 'netdriver' of https://github.com/lf-lang/reactor-c into netdriver

Jakio815 committed:
All tests except FeedbackDelay5.lf works

Jakio815 committed:
Formatting federate.c

Jakio815 committed:
Fixed federate compile

Jakio815 committed:
Fix merge conflicts

Jakio815 committed:
RTI now compiles selectively by COMM_TYPE keyword

Jakio815 committed:
Add cmake to reactor-c cmake

Jakio815 committed:
Start sst support

chanijjani committed:
Merge branch 'netdriver' of github.com:lf-lang/reactor-c into netdriver

chanijjani committed:
Enable MQTT build

Jakio815 committed:
Merge branch 'netdriver' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Change reactor-c CMakelists.txt, Add network apis.

Jakio815 committed:
Comment out federated/network/CMakeLists.txt

Jakio815 committed:
Change path for core/Cmakelists.txt

chanijjani committed:
Fix federate build

chanijjani committed:
Merge branch 'netdriver' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Change netdriver to drv in arguments

Jakio815 committed:
Move create server to lf_socket support.c && Move socket_open to socket_common.c

Jakio815 committed:
Revert commented out parts

Jakio815 committed:
Make sst build work

Jakio815 committed:
Remove server_type_t temporarily for compile...

Jakio815 committed:
Add create_server and establish_communication_session for SST

Jakio815 committed:
Remove gitmodules and sst-c-api

chanijjani committed:
Fix errors during lfc-dev build

chanijjani committed:
Merge branch 'netdriver' of github.com:lf-lang/reactor-c into netdriver

chanijjani commented:
@Jakio815 I updated the reactor-c code so that the `lfc-dev` build with `SimpleFederated.lf` and the `RTI` build are done successfully. I would appreciate it if you could check it when you have time.
When implementing the `comm-type` option in the `SimpleFederated.lf` finishes, I'll update the netdriver branch in the lingua-franca repository.

Jakio815 committed:
Minor fix on sst support.c

Jakio815 committed:
Remove comments

Jakio815 committed:
Fix handle clock sync warnings and errors.

Jakio815 committed:
Apply formatting

Jakio815 committed:
Re-link lov-level-platform-impl to lf-network-api

Jakio815 committed:
Add skeleton for netdrv_connect, peek_from_netdrv at lf_socket_support.c & mqtt support.c

chanijjani committed:
Enable comm-type target property

chanijjani committed:
Merge branch 'netdriver' of github.com:lf-lang/reactor-c into netdriver

chanijjani committed:
Merge branch 'netdriver' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Revert reactor-c/core/CMakelists.

Jakio815 committed:
Fix handle tagged_messages

Jakio815 committed:
Remove unused header includes in main.c

Jakio815 committed:
Merge branch 'netdriver' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Minor fix

Jakio815 committed:
Add SST configs

Jakio815 committed:
Add set_specified_port()

Jakio815 committed:
Minor fix

Jakio815 committed:
Change logic at lf_connect_to_rti

Jakio815 committed:
Minor fix

Jakio815 committed:
Update sst support. establish comm,  netdrv connect. Still needs fix

Jakio815 committed:
Minor fix

Jakio815 committed:
Fix fed config

Jakio815 committed:
Fix fed_config

Jakio815 committed:
Fix netdrv->open to get federtate ID

Jakio815 committed:
Change logic for federate's connect.

Jakio815 committed:
Change sst_open, get federate ID. && Add read and write functions using SST.

Jakio815 committed:
Minor fix

hokeun committed:
Fix build issues on Mac

Jakio815 committed:
" Fix header includes"

Jakio815 committed:
Add #include <unistd.h>

Jakio815 committed:
Fixed to read on mesasge based, not going over the next message.

Jakio815 committed:
Minor fix

hokeun committed:
Fix call to undeclared function write; ISO C99 and later do not support implicit function declarations for write and read functions

hokeun committed:
Fix fatal error: sst-c-api/c_api.h file not found by adding include_directories(/usr/local/include)

hokeun committed:
Fix fatal error: sst-c-api/c_api.h file not found b
y adding include_directories(/usr/local/include)

hokeun committed:
Fix compiler warnings

hokeun committed:
Fix error: passing 'unsigned char *' to parameter of type 'char *' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Werror,-Wpointer-sign]

Jakio815 committed:
Minor fix - remove commented out functions

Jakio815 committed:
Get federate's sst config path as command line

Jakio815 committed:
Change federate config name

Jakio815 committed:
Add --sst options to federate

Jakio815 committed:
Add SSTskeleton.config

Jakio815 committed:
Minor fix -remove unneeded configs.

Jakio815 committed:
Warning fix

Jakio815 committed:
Add RTI options to get -sst as sstconfig path.

hokeun committed:
Fix a function not returnning anything

hokeun committed:
Revert wrong commit

Jakio815 committed:
Add mqtt netdrv_init()

Jakio815 committed:
Change netdrv_init() to get federate id and federation id.

Jakio815 committed:
Remove type/ header from each support. files.

Jakio815 committed:
Add each support.h headers to the c files.

Jakio815 committed:
Add address, qos, timeout values.

Jakio815 committed:
Move qos address, timeout values to .c file.

Jakio815 committed:
Add create server and read to mqtt support.

Jakio815 committed:
Minor fix

Jakio815 committed:
Implement mqtt support. Almost working....

Jakio815 committed:
Implement read, write, connect... so on...

Jakio815 committed:
Add mqtt join and ack messages

Jakio815 committed:
Fix bugs and add mqtt join and ack protocols.

Jakio815 committed:
Remove function pointers of open and close.

Jakio815 committed:
Reorganize code and change some names.

Jakio815 committed:
Add netdriver.c to take out common functions of the read and write close and fail on errors.

Jakio815 committed:
Reorganize sst support.c

Jakio815 committed:
Reorganize

Jakio815 committed:
Reorganize

Jakio815 committed:
Minor fix

Jakio815 committed:
Fix mqtt support. Remove base64 encoding decoding and it's libraries.

Jakio815 committed:
Fix read to return 0 on connection lost.

Jakio815 committed:
Fix rti config

Jakio815 committed:
Add skeleton for openssl_required definition.

Jakio815 committed:
Add temporary configs.

Jakio815 committed:
Make a common connect to socket function.

Jakio815 committed:
Minor fix

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Fix conflicts

Jakio815 committed:
Fix unsused variable errors.

Jakio815 committed:
Add resign message

Jakio815 committed:
FIx sst read

Jakio815 committed:
Remove library

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Fix to return code

Jakio815 committed:
Fix to get server_type_t when create_server()

Jakio815 committed:
Minor fix

Jakio815 committed:
Add initialize_common_netdrv() to not repeat code && Apply server_type_t

Jakio815 committed:
Add initialize_common_netdrv

Jakio815 committed:
Minor fix on netdriver CMake.

Jakio815 committed:
Add send_address_advertisement_to_RTI && minor formatting.

Jakio815 committed:
Minor comments and formatting.

Jakio815 committed:
Fix declaration of send_address_advertisement_to_RTI

Jakio815 committed:
Change function names to connector and listener

Jakio815 committed:
Move send_address_advertisement_to_RTI() back to federate.c

Jakio815 committed:
Change multiple term changes.

federate_id-> my_federate_id, my_netdrv->listener_netdrv, ret_netdrv->connector_netdrv

Jakio815 committed:
Fix connect_to_socket with timeout time and better errors

Jakio815 committed:
Fix federate ID to uint16_t

Jakio815 committed:
Create function create_topic_federation_id_A_to_B to make a topic name with the federateID and RTI name.

Jakio815 committed:
Add set_MQTT_target_ID.

Jakio815 committed:
Add stdint.h to mqtt_support.h

Jakio815 committed:
Minor fixes.

Jakio815 committed:
Set back federate ID to int

Jakio815 committed:
Fix bugs, add comments.

Jakio815 committed:
Make connect retries for MQTT connect()

Jakio815 committed:
Minor comments.

Jakio815 committed:
Minor change on MQTT connect.

Jakio815 committed:
Minor fix

Jakio815 committed:
Add retries for connect and subscribe

Jakio815 committed:
Add retry for the total logic itself.

Jakio815 committed:
Take the read outside from the while loop

Jakio815 committed:
Fix to establish communication first, then do the fedID etc.. hanshake.

Jakio815 committed:
Remove MQTT connect with retries.

Jakio815 committed:
Change message name from MQTT_RTI_RESIGNED to MQTT_RESIGNED.

Jakio815 committed:
Better close_netdrv() on MQTT support && Minor fix

Jakio815 committed:
Update comments from socket to netdriver

Jakio815 committed:
Update create_TCP_server for better port handling.

Jakio815 committed:
Minor fix.

Jakio815 committed:
Add MQTTClient_receive to retry.

Jakio815 committed:
Minor fixes. Enable handling resign.

Jakio815 committed:
Fix Clock-sync to work on MQTT.

Jakio815 committed:
Fix many comments and bugs.

Jakio815 committed:
Add better logs && Fix names for conflicts  on topic and ClientID names.

Jakio815 committed:
Add checking for drv is open.

Jakio815 committed:
Fix checking if netdrv to rti is not NULL

Jakio815 committed:
Move netdrv_mutex to netdriver.c

Jakio815 committed:
Add LF_MUTEX_INIT(&netdrv_mutex); for also the rti use the same mutex for netdrivers.

Jakio815 committed:
Minor fix

Jakio815 committed:
Remove log

Jakio815 committed:
Add mutex and error handling. Still getting segfaults on termination.

Jakio815 committed:
Reveal mutex locks..

Jakio815 committed:
Removing close netdrivers on netdrv_close_on_error

Jakio815 committed:
Move _lf_termination_executed to netdriver.c for checking the status during every read.

Jakio815 committed:
Remove old configs.

Jakio815 committed:
Add comments.

Jakio815 committed:
Apply Erling's fix on clearing out chunks.

Jakio815 committed:
Checked buffer length for read

Jakio815 committed:
Change bytes_read to 0 for close() situations.

Jakio815 committed:
Minor fix

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Fix RTI build

Jakio815 committed:
Merge branch 'main' of github.com:lf-lang/reactor-c into netdriver

Jakio815 committed:
Temp commit.

Jakio815 committed:
Minor fix.

Jakio815 commented on include/core/federated/federate.h:
```diff
@@ -345,7 +340,7 @@ int lf_send_message(int message_type, unsigned short port, unsigned short federa
  * @see MSG_TYPE_NEIGHBOR_STRUCTURE in net_common.h
  */
-void lf_send_neighbor_structure_to_RTI(int);
+void lf_send_neighbor_structure_to_RTI(netdrv_t* netdrv);
```
Check.
